### PR TITLE
Add incremental updates for SSGPR

### PR DESCRIPTION
**Pull request by kazuotani14**

### CODE REVIEW

File: src/limbo/experimental/model/ssgpr.hpp [hunk 1]
```diff
@@ -32,10 +33,10 @@ namespace limbo {
                 {
                     assert(_dim_in > 0 && _dim_out > 0);
 
-                    int n = _mapping.dim_out();
-                    _L = _sigma_n * _sigma_n * Eigen::MatrixXd::Identity(n, n);
-                    _B = Eigen::MatrixXd::Zero(n, _dim_out);
-                    _W = Eigen::MatrixXd::Zero(n, _dim_out);
+                    _dim_mapping = _mapping.dim_out();
+                    _L = _sigma_n * Eigen::MatrixXd::Identity(_dim_mapping, _dim_mapping);
```
Comment (costashatz) [inline src/limbo/experimental/model/ssgpr.hpp]: Why did you remove the _sigma_n^2? I don't really mind, but just to know..

File: src/limbo/experimental/model/ssgpr.hpp [hunk 2]
```diff
@@ -169,8 +170,59 @@ namespace limbo {
...
+                this->_compute_obs_mean();
+
+                _Phi.conservativeResize(_Phi.rows()+1, _mapping.dim_out());
+                Eigen::VectorXd sample_phi = _mapping.evaluate(sample);
+                _Phi.bottomRows<1>() = sample_phi;
+
+                // this->_compute_incremental_kernel(); 
+
+                // Make Rtilde - [R, sample_phi]
+                Eigen::MatrixXd Rtilde(_dim_mapping+1, _dim_mapping);
```
Comment (costashatz) [inline src/limbo/experimental/model/ssgpr.hpp]: Can't we avoid create a whole new matrix and just re-size `_L`?

File: src/limbo/experimental/model/ssgpr.hpp [hunk 3]
```diff
@@ -169,8 +170,59 @@ namespace limbo {
...
+                Rtilde.topRows(_dim_mapping) = _L.transpose();
+                Rtilde.bottomRows<1>() = sample_phi;
+
+                // Note: the Gijsberts paper uses Givens rotations. Is that a bit more efficient?
+                Eigen::HouseholderQR<Eigen::MatrixXd> qr_res2 = Rtilde.householderQr();
+                _L = qr_res2.matrixQR().topRows(_dim_mapping).transpose();
+
+                _B += sample_phi * (observation -  _mean_function(sample, *this)).transpose();
+
+                // TODO Adapt to incremental? cost is O(_dim_mapping^2) which isn't that bad
```
Comment (costashatz) [inline src/limbo/experimental/model/ssgpr.hpp]: I do not remember the exact math for this. Is it easy to add? If not, I think it'll be ok like this..

Comment (kazuotani14) [reply src/limbo/experimental/model/ssgpr.hpp]: I did this to make the update step analogous to what's done in batch learning: 

```
for (int i = 0; i < _mean_vector.rows(); i++)
    _mean_vector.row(i) = _mean_function(_samples[i], *this);
_obs_mean = _observations - _mean_vector;
...
_B = _Phi.transpose() * _obs_mean;
```

File: src/limbo/experimental/model/ssgpr.hpp (hunk 1 above)
Comment (kazuotani14) [reply src/limbo/experimental/model/ssgpr.hpp]: This is from I-SSGPR paper section 3.3, where they say R0 should be the cholesky factor of sigma_n^2 eye(D).

File: src/limbo/experimental/model/ssgpr.hpp (hunk 2 above)
Comment (kazuotani14) [reply src/limbo/experimental/model/ssgpr.hpp]: Do you mean that we should resize `_L` to `(_dim_mapping+1, _dim_mapping)`, do some operations, then resize back to `(_dim_mapping, _dim_mapping)`? Is this better than having a pre-allocated member variable to fill with values at each update?

File: src/limbo/experimental/model/ssgpr.hpp (hunk 1 above)
Comment (costashatz) [reply src/limbo/experimental/model/ssgpr.hpp]: > This is from I-SSGPR paper section 3.3, where they say R0 should be the cholesky factor of sigma_n^2 eye(D).

True dat..

File: src/limbo/experimental/model/ssgpr.hpp (hunk 2 above)
Comment (costashatz) [reply src/limbo/experimental/model/ssgpr.hpp]: > Do you mean that we should resize _L to (_dim_mapping+1, _dim_mapping), do some operations, then resize back to (_dim_mapping, _dim_mapping)? Is this better than having a pre-allocated member variable to fill with values at each update?

What I mean is something like the following (correct me if I'm wrong and this would not work):

```cpp
// Make Rtilde - [R, sample_phi]
_L = _L.transpose();
_L.conservativeResize(_dim_mapping+1, _dim_mapping);
_L.bottomRows<1>() = sample_phi;

// Note: the Gijsberts paper uses Givens rotations. Is that a bit more efficient?
Eigen::HouseholderQR<Eigen::MatrixXd> qr_res2 = _L.householderQr();
_L = qr_res2.matrixQR().topRows(_dim_mapping).transpose();
```

File: src/limbo/experimental/model/ssgpr.hpp (hunk 3 above)
Comment (costashatz) [reply src/limbo/experimental/model/ssgpr.hpp]: OK..

File: src/limbo/experimental/model/ssgpr.hpp (hunk 2 above)
Comment (kazuotani14) [reply src/limbo/experimental/model/ssgpr.hpp]: You're right, that should work. I added the change.

**Comment by costashatz** [comment]
**Looks good! Merging!**

Thanks again @kazuotani14!
